### PR TITLE
Huge leak fixes proper

### PR DESCRIPTION
**Pull request by Therzok**

### CODE REVIEW

**Comment by Therzok** [comment]
@mkrueger please take a look at https://github.com/mono/monodevelop/pull/2313/commits/ec10b4c258d361fec347262d26dc378cdfed2930

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/Document.cs [hunk 1]
```diff
@@ -674,7 +674,7 @@ void InitializeEditor ()
 
 				if (Editor.IsInAtomicUndo) {
 					wasEdited = true;
-				} else {
+				} else if (!wasEdited) {
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/Document.cs]: @mkrueger is this correct?

Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/Document.cs]: No, it's not.

Comment (mkrueger) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/Document.cs]: No it's not :)

File: main/src/core/Mono.TextEditor.Shared/Mono.TextEditor/Document/TextDocument.cs [hunk 2]
```diff
@@ -222,6 +222,14 @@ protected void Initialize()
 			this.diffTracker.SetTrackDocument(this);
 		}
 
+		public void Dispose()
+		{
+			this.TextBuffer.Changed -= this.OnTextBufferChanged;
+			this.TextBuffer.ContentTypeChanged -= this.OnTextBufferContentTypeChanged;
+			this.TextBuffer.Properties.RemoveProperty(typeof(ITextDocument));
+			this.VsTextDocument.FileActionOccurred -= this.OnTextDocumentFileActionOccured;
```
Comment (mkrueger) [inline main/src/core/Mono.TextEditor.Shared/Mono.TextEditor/Document/TextDocument.cs]: Interesting that this is required. What about the other handlers ?
TextChanging += HandleSplitterLineSegmentTreeLineRemoved;
foldSegmentTree.tree.NodeRemoved += HandleFoldSegmentTreetreeNodeRemoved;

Or the reference in diffTracker ?

In theory all these handlers should be disposed automatically - is something keping textbuffer & vstextdocument alive ?

Comment (Therzok) [inline main/src/core/Mono.TextEditor.Shared/Mono.TextEditor/Document/TextDocument.cs]: Yes, it's NavigationHistoryService. We keep the textbuffer alive, which keeps the textdocument in the property bag and the vstextdocument via the handler.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CSharpCompletionTextEditorExtension.cs [hunk 3]
```diff
@@ -242,27 +237,8 @@ void HandleDocumentParsed (object sender, EventArgs e)
...
-				}
-			});
-		}
-
-		void CancelParsedDocumentUpdate ()
-		{
-			documentParsedTokenSrc.Cancel ();
-			documentParsedTokenSrc = new CancellationTokenSource ();
+				return;
+			TypeSegmentTreeUpdated?.Invoke (this, EventArgs.Empty);
```
Comment (decriptor) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CSharpCompletionTextEditorExtension.cs]: Doesn't the `return;` above make this line useless?

**Comment by Therzok** [comment]
@slluis we'll want these changes in master too. They are the same as d15-2 commits, except the dockitemtitletab change and just computed highlighted line.

**Comment by Therzok** [comment]
Superseded by #2346

**Pull request closed without merging**